### PR TITLE
[mono] Stop exporting ICU symbols from Mono

### PR DESCRIPTION
**Pull request by matouskozak**
## Customer Impact

- [X] Customer reported
- [X] Found internally

When running against the latest XCode, the iOS 17.4 simulator causes our apps to crash due to a symbols clash between their version of ICU and our version. Prior to this change, by default we had been exporting all our symbols and been lucky due to the ICU versions matching. To fix, we hide our symbols via `-hidden-lx` so that there is no longer a clash.

NOTE: This does not impact .NET 9 as we have moved away from shipping our own version of ICU on iOS.

## Regression
- [X] Yes
- [ ] No

## Testing
We verified the fix thru manually copying new versions of `libmonosgen-2.0.dylib` and `libSystem.Globalization.Native.dylib` to previously crashing app and re-running it on the simulator. Currently, we are unable to verify this E2E by using a template iOS app with custom runtime due to issues with trampolines.

## Risk
Low, but since we were unable to verify E2E with a template app, there may be more churn required if the iOS SDK team finds any issues.

---

Contributed: @ivanpovazan 
fyi: @vitek-karas @akoeplinger @steveisok @dalexsoto @mkhamoyan

### CODE REVIEW

**Comment by matouskozak** [comment]
We verified the fix thru manually copying new versions of `libmonosgen-2.0.dylib` and `libSystem.Globalization.Native.dylib` to previously crashing app and re-running it on the simulator. Currently, we are unable to verify this E2E by using a template iOS app with custom runtime due to following issue with trampolines:
```
Thread 0 Crashed::  Dispatch queue: com.apple.main-thread
0   libsystem_platform.dylib      	       0x10ec25828 _platform_memset_pattern16 + 296
1   libmonosgen-2.0.dylib         	       0x1113f0fa8 mono_arch_create_sdb_trampoline + 76
2   libmonosgen-2.0.dylib         	       0x1113a9134 mini_get_breakpoint_trampoline + 100
3   libmonosgen-2.0.dylib         	       0x1113da7a8 mono_arch_init + 44
4   libmonosgen-2.0.dylib         	       0x1113147f8 mini_init + 848
5   libmonosgen-2.0.dylib         	       0x11136b7d0 mono_jit_init_version + 20
6   libxamarin-dotnet-debug.dylib 	       0x110095550 xamarin_bridge_initialize + 216
7   libxamarin-dotnet-debug.dylib 	       0x11009b8e4 xamarin_main + 1780
8   simulator_ICU_issue           	       0x106b2b868 main + 64 (main.arm64.mm:409)
9   dyld_sim                      	       0x10f209544 start_sim + 20
10  dyld                          	       0x10ed060e0 start + 2360

Thread 1:
0   libsystem_pthread.dylib       	       0x10fbe66dc start_wqthread + 
```
that probably comes from https://github.com/dotnet/runtime/blob/e7e3dee0a50e1780c889d63797089acc12e564da/src/mono/mono/mini/mini-trampolines.c#L1666-L1677 where the `mono_ee_features.use_aot_trampolines` gets evaluated to false. @rolfbjarne Do you have any idea on what could be the issue or another working approach on how to tests this with custom runtime and iOS template app?

File: src/native/libs/System.Globalization.Native/CMakeLists.txt [hunk 1]
```diff
@@ -68,6 +68,13 @@ if (DEFINED CMAKE_ICU_DIR)
     include_directories(${CMAKE_ICU_DIR}/include)
     link_libraries(${CMAKE_ICU_DIR}/lib/libicuuc.a ${CMAKE_ICU_DIR}/lib/libicui18n.a ${CMAKE_ICU_DIR}/lib/libicudata.a)
     link_libraries(stdc++)
+
+    if (CLR_CMAKE_TARGET_IOS)
```
Comment (akoeplinger) [inline src/native/libs/System.Globalization.Native/CMakeLists.txt:72]:
```suggestion
    if (CLR_CMAKE_TARGET_MACCATALYST OR CLR_CMAKE_TARGET_IOS OR CLR_CMAKE_TARGET_TVOS)
```

File: src/mono/mono/mini/CMakeLists.txt [hunk 2]
```diff
@@ -421,7 +421,13 @@ if(NOT DISABLE_SHARED_LIBS)
     # to avoid a conflict we rename the import library with the .import.lib suffix
     set_target_properties(monosgen-shared PROPERTIES IMPORT_SUFFIX ".import.lib")
   endif()
-  target_link_libraries(monosgen-shared PRIVATE ${OS_LIBS} ${LLVM_LIBS} ${ICU_LIBS} ${Z_LIBS})
+  target_link_libraries(monosgen-shared PRIVATE ${OS_LIBS} ${LLVM_LIBS} ${Z_LIBS})
+  if (CLR_CMAKE_TARGET_IOS)
```
Comment (akoeplinger) [inline src/mono/mono/mini/CMakeLists.txt:425]:
```suggestion
  if (CLR_CMAKE_TARGET_MACCATALYST OR CLR_CMAKE_TARGET_IOS OR CLR_CMAKE_TARGET_TVOS)
```

**Comment by rolfbjarne** [comment]
> @rolfbjarne Do you have any idea on what could be the issue or another working approach on how to tests this with custom runtime and iOS template app?

I wonder if the dylib needs to be signed?

Do you get the same error if you copy:

1. An unmodified dylib from a published NuGet.
2. A lcoally built dylib, from the same hash has the published NuGet without any source code changes.

An alternative might be to try in the x64 simulator.

**Comment by ivanpovazan** [comment]
A note regarding E2E testing and:
```
Thread 0 Crashed::  Dispatch queue: com.apple.main-thread
0   libsystem_platform.dylib      	       0x10ec25828 _platform_memset_pattern16 + 296
1   libmonosgen-2.0.dylib         	       0x1113f0fa8 mono_arch_create_sdb_trampoline + 76
2   libmonosgen-2.0.dylib         	       0x1113a9134 mini_get_breakpoint_trampoline + 100
3   libmonosgen-2.0.dylib         	       0x1113da7a8 mono_arch_init + 44
4   libmonosgen-2.0.dylib         	       0x1113147f8 mini_init + 848
5   libmonosgen-2.0.dylib         	       0x11136b7d0 mono_jit_init_version + 20
6   libxamarin-dotnet-debug.dylib 	       0x110095550 xamarin_bridge_initialize + 216
7   libxamarin-dotnet-debug.dylib 	       0x11009b8e4 xamarin_main + 1780
8   simulator_ICU_issue           	       0x106b2b868 main + 64 (main.arm64.mm:409)
9   dyld_sim                      	       0x10f209544 start_sim + 20
10  dyld                          	       0x10ed060e0 start + 2360

Thread 1:
0   libsystem_pthread.dylib       	       0x10fbe66dc start_wqthread +
```

We are getting this even if we build the runtime locally and configure the project file to reference the locally built runtime like:
```xml
  <PropertyGroup Condition="$(LocalBuild) == 'true'">
    <RestoreAdditionalProjectSources>/Users/ivan/repos/runtime-mono/artifacts/packages/Release/Shipping</RestoreAdditionalProjectSources>
  </PropertyGroup>

  <ItemGroup Condition="$(LocalBuild) == 'true'">
    <FrameworkReference Update="Microsoft.NETCore.App"  RuntimeFrameworkVersion="8.0.3-dev" />
  </ItemGroup>
```

However, if we build the app and manually copy `libmonosgen-2.0.dylib` and `libSystem.Globalization.Native.dylib` over to the built app bundle and run it on the simulator, the `mono_arch_create_sdb_trampoline` problem is not triggered.

---
I remember @lambdageek experienced the same problem a while back: https://github.com/dotnet/runtime/issues/93860#issuecomment-1808659507

**Comment by rolfbjarne** [comment]
> Thread 0 Crashed::  Dispatch queue: com.apple.main-thread
> 0   libsystem_platform.dylib      	       0x10ec25828 _platform_memset_pattern16 + 296
> 1   libmonosgen-2.0.dylib         	       0x1113f0fa8 mono_arch_create_sdb_trampoline + 76

IIRC it's related to creating executable + writable memory (or failing to do so).

You get the same crash on Mac Catalyst if you enable the sandbox, but don't add the jit entitlement.

**Comment by ivanpovazan** [comment]
> > Thread 0 Crashed::  Dispatch queue: com.apple.main-thread
> > 0   libsystem_platform.dylib      	       0x10ec25828 _platform_memset_pattern16 + 296
> > 1   libmonosgen-2.0.dylib         	       0x1113f0fa8 mono_arch_create_sdb_trampoline + 76
> 
> IIRC it's related to creating executable + writable memory (or failing to do so).
> 
> You get the same crash on Mac Catalyst if you enable the sandbox, but don't add the jit entitlement.

If I understand the flow correctly, we shouldn't be getting to the `mono_arch_create_sdb_trampoline` at all.

For the app we generate:
```
void xamarin_setup_impl ()
{
	mono_jit_set_aot_mode (MONO_AOT_MODE_FULL);
```
which should be invoked from xamarin's main:
https://github.com/xamarin/xamarin-macios/blob/58e382efb7abdb9cf88900bd8e87f27f7e586a51/runtime/monotouch-main.m#L276

that further gets here:
https://github.com/dotnet/runtime/blob/80c0df6bca63c09889411035395ca50dc7a4b0bc/src/mono/mono/mini/driver.c#L2910-L2932

and finally here:
https://github.com/dotnet/runtime/blob/80c0df6bca63c09889411035395ca50dc7a4b0bc/src/mono/mono/mini/driver.c#L2852-L2865

which sets the: `mono_ee_features.use_aot_trampolines = TRUE;` and should guard against trying to create executable code (sdb trampolines) here:
https://github.com/dotnet/runtime/blob/e7e3dee0a50e1780c889d63797089acc12e564da/src/mono/mono/mini/mini-trampolines.c#L1666-L1677

Will have to set up debug builds of everything I suppose .. :)

**Review by akoeplinger — APPROVED** [review]

**Comment by matouskozak** [comment]
This fix probably won't be enough. 

I was able to re-produce the original issue (https://github.com/dotnet/runtime/issues/98941) without Xamarin by trying the Hybrid Globalization tests https://github.com/dotnet/runtime/blob/release/8.0-staging/src/libraries/System.Globalization/tests/Hybrid/System.Globalization.IOS.Tests.csproj. The error message is the same 
```
System.Globalization.IOS.Tests(42002,0x16df7f000) malloc: *** error for object 0x16df7a695: pointer being freed was not allocated
System.Globalization.IOS.Tests(42002,0x16df7f000) malloc: *** set a breakpoint in malloc_error_break to debug
```
and it also only fails on the newest iOS simulator (17.2 works fine) + it works on the .NET 9 main.

Do you know if we possibly export ICU symbols somewhere else in the Mono Runtime? @akoeplinger  @mkhamoyan 
Or any other idea what could be causing this?

**Comment by rolfbjarne** [comment]
> Do you know if we possibly export ICU symbols somewhere else in the Mono Runtime?

You can run `nm` on Mono's libraries to check which symbols they export.

**Comment by akoeplinger** [comment]
The code in src/mono/mono/mini/CMakeLists.txt _should_ have fixed it but yeah check with nm.

**Comment by matouskozak** [comment]
> The code in src/mono/mono/mini/CMakeLists.txt _should_ have fixed it but yeah check with nm.

I double checked that and both `libSystem.Globalization.Native.dylib` and `libmonosgen-2.0.dylib` don't contain the ICU symbols, only our APIs `_GlobalizationNative_*`.

**Comment by am11** [comment]
I think https://github.com/dotnet/runtime/pull/98495 has superseded this.

**Comment by vitek-karas** [comment]
@am11 - that's true for .NET 9, but this change is trying to fix this for .NET 8.

**Review by akoeplinger — APPROVED** [review]

**Review by steveisok — APPROVED** [review]
LGTM! @matouskozak library mode should be ok because we only export symbols declared in `UnmanagedCallersOnly` and hide everything else.

**Review by jeffschwMSFT — APPROVED** [review]
approved. we will take for consideration in 8.0.x

**Review by rolfbjarne — APPROVED** [review]

**Pull request merged**